intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+
+@torch.jit.script
+def fn(x):
+    return x
+
+catch_error(lambda: fn(imgs_named))
+
+######################################################################
+# As a workaround, please drop names via `tensor = tensor.rename(None)`
+# before using anything that does not yet support named tensors.
+#
+# Longer example: Multi-headed attention
+# --------------------------------------
+#
+# Now we'll go through a complete example of implementing a common advanced
+# PyTorch ``nn.Module``: multi-headed attention. We assume the reader is already
+# familiar with multi-headed attention; for a refresher, check out
+# `this explanation <http://jalammar.github.io/illustrated-transformer/>`_.
+#
+# We adapt the implementation of multi-headed attention from
+# `ParlAI <https://github.com/facebookresearch/ParlAI>`_; specifically
+# `here <https://github.com/facebookresearch/ParlAI/blob/f7db35cba3f3faf6097b3e6b208442cd564783d9/parlai/agents/transformer/modules.py#L907>`_.
+# Read through the code at that example; then, compare with the code below,
+# noting that there are four places labeled (I), (II), (III), and (IV), where
+# using named tensors enables more readable code; we will dive into each of these
+# after the code block.
+
+import torch.nn as nn
+import torch.nn.functional as F
+import math
+
+class MultiHeadAttention(nn.Module):
+    def __init__(self, n_heads, dim, dropout=0):
+        super(MultiHeadAttention, self).__init__()
+        self.n_heads = n_heads
+        self.dim = dim
+
+        self.attn_dropout = nn.Dropout(p=dropout)
+        self.q_lin = nn.Linear(dim, dim)
+        self.k_lin = nn.Linear(dim, dim)
+        self.v_lin = nn.Linear(dim, dim)
+        nn.init.xavier_normal_(self.q_lin.weight)
+        nn.init.xavier_normal_(self.k_lin.weight)
+        nn.init.xavier_normal_(self.v_lin.weight)
+        self.out_lin = nn.Linear(dim, dim)
+        nn.init.xavier_normal_(self.out_lin.weight)
+
+    def forward(self, query, key=None, value=None, mask=None):
+        # (I)
+        query = query.refine_names(..., 'T', 'D')
+        self_attn = key is None and value is None
+        if self_attn:
+            mask = mask.refine_names(..., 'T')
+        else:
+            mask = mask.refine_names(..., 'T', 'T_key')  # enc attn
+
+        dim = query.size('D')
+        assert dim == self.dim, \
+            f'Dimensions do not match: {embedding_dim} query vs {self.dim} configured'
+        assert mask is not None, 'Mask is None, please specify a mask'
+        n_heads = self.n_heads
+        dim_per_head = dim // n_heads
+        scale = math.sqrt(dim_per_head)
+
+        # (II)
+        def prepare_head(tensor):
+            tensor = tensor.refine_names(..., 'T', 'D')
+            return (tensor.unflatten('D', [('H', n_heads), ('D_head', dim_per_head)])
+                          .align_to(..., 'H', 'T', 'D_head'))
+
+        if self_attn:
+            key = value = query
+        elif value is None:

```

Comment:
should probably remove the value is None now... and `assert value is None`

Reply:
Here, `value is None` is `True`, so we want to run the line `key = value = query`, right?

So we want to remove the `elif` block? That's what I'll do in the next version.